### PR TITLE
Match ultra boots

### PR DESCRIPTION
**Pull request by ThreeMustyFears**

### CODE REVIEW

File: src/world/action/ultra_boots.c [hunk 1]
```diff
@@ -1,5 +1,200 @@
...
+    case 2:
+        temp_f20 = integrate_gravity();
+        playerStatus->position.y = player_check_collision_below(temp_f20, &sp10);
+        if (temp_f20 < -100.0f) {
+            playerStatus->gravityIntegrator[3] = 0.0f;
+            playerStatus->gravityIntegrator[2] = 0.0f;
+            playerStatus->gravityIntegrator[1] = 0.0f;
+            playerStatus->gravityIntegrator[0] = -100.0f;
+        }
+        if (sp10 >= 0) {
```
Comment (ethteck) [inline src/world/action/ultra_boots.c:112]: The control flow from here on is a bit hard to figure out with all the early returns. I think you might be able to if/else a lot of this

**Review by ethteck — CHANGES_REQUESTED** [review]
Nice job on this! Left some style comments.

File: src/world/action/ultra_boots.c [hunk 2]
```diff
@@ -1,5 +1,200 @@
...
+    f32 temp_f20;
+    s32 temp_a0;
+    s32 temp_v0_2;
+    s32 temp_v1_2;
+    u32 temp_v0;
+    f32 phi_f4;
+    s32 sp12;
+    s32 sp10;
+
+    if (playerStatus->flags < 0) {
```
Comment (nanaian) [inline src/world/action/ultra_boots.c:30]:
```suggestion
    if (playerStatus->flags & (1 << 31)) {
```

File: src/world/action/ultra_boots.c [hunk 3]
```diff
@@ -1,5 +1,200 @@
...
+        }
+        if (sp10 >= 0) {
+            if (collisionStatus->currentFloor & 0x4000) {
+                temp_v0 = get_entity_type(collisionStatus->currentFloor);
+                if (temp_v0 - 0x2E < 2) {
+                    playerStatus->flags &= ~0x00020008;
+                    set_action_state(ACTION_STATE_LAND);
+                    return;
+                }
+                if (temp_v0 - 7 < 2) {
```
Comment (nanaian) [inline src/world/action/ultra_boots.c:120]:
```suggestion
                if (temp_v0 == 7 || temp_v0 = 8) {
```

File: src/world/action/ultra_boots.c [hunk 4]
```diff
@@ -1,5 +1,200 @@
...
+        if (temp_f20 < -100.0f) {
+            playerStatus->gravityIntegrator[3] = 0.0f;
+            playerStatus->gravityIntegrator[2] = 0.0f;
+            playerStatus->gravityIntegrator[1] = 0.0f;
+            playerStatus->gravityIntegrator[0] = -100.0f;
+        }
+        if (sp10 >= 0) {
+            if (collisionStatus->currentFloor & 0x4000) {
+                temp_v0 = get_entity_type(collisionStatus->currentFloor);
+                if (temp_v0 - 0x2E < 2) {
```
Comment (nanaian) [inline src/world/action/ultra_boots.c:115]:
```suggestion
                if (temp_v0 == 0x2E || temp_v0 == 0x2F) {
```

File: src/world/action/ultra_boots.c [hunk 5]
```diff
@@ -1,5 +1,200 @@
...
+                    start_rumble(0x100, 0x32);
+
+                    tempStruct = &D_8015A578;
+                    tempStruct->unk_00 = 1;
+                    tempStruct->unk_08 = playerStatus->position.y;
+                    playerStatus->flags |= 0x400;
+                    return;
+                }
+            }
+            temp_v1_2 = get_collider_type_by_id(sp10) & 0xFF;
```
Comment (nanaian) [inline src/world/action/ultra_boots.c:134]: Change this variable's type to `s8` to get rid of this explicit cast

**Review by nanaian — CHANGES_REQUESTED** [review]

File: src/world/action/ultra_boots.c (hunk 5 above)
Comment (ethteck) [reply src/world/action/ultra_boots.c:134]: that'd be u8

File: src/world/action/ultra_boots.c (hunk 4 above)
Comment (ThreeMustyFears) [reply src/world/action/ultra_boots.c:115]: This is similar to the previous one so I was just wondering if you want to keep it as 0x2E, and 0x2F or if it should be changed 46, and 47 respectively

File: src/world/action/ultra_boots.c [hunk 6]
```diff
@@ -63,25 +61,29 @@ void func_802B6000_E26710(void) {
...
-        if (sp10 >= 0 && collisionStatus->currentFloor & 0x4000 && (get_entity_type(collisionStatus->currentFloor) - 7 < 2)) {
-            currentFloorEntity = get_entity_by_index(collisionStatus->currentFloor);
-            currentFloorEntity->collisionFlags |= 1;
-            disable_player_input();
-            playerStatus->fallState = 0xB;
-        } else {
+        case 0:
+            temp_f20 = integrate_gravity();
+            playerStatus->position.y = player_check_collision_below(temp_f20, &sp10);
+            if (sp10 >= 0 && collisionStatus->currentFloor & 0x4000 ) {
```
Comment (nanaian) [inline src/world/action/ultra_boots.c:77]: There's a random space here before the closing parenthesis

File: src/world/action/ultra_boots.c (hunk 4 above)
Comment (nanaian) [reply src/world/action/ultra_boots.c]: I don't mind; I'm sure we'll be adding an enum for these soon

**Review by nanaian — APPROVED** [review]
LGTM, thanks for making changes

**Comment by ethteck** [comment]
Thanks! Merging

**Pull request merged**